packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.kt

```diff
@@ -237,12 +237,35 @@ class UpdatesModule(
                 return
               }
 
+              var shouldLaunch = false
               if (updatesServiceLocal.selectionPolicy.shouldLoadNewUpdate(
                   updateManifest.updateEntity,
                   launchedUpdate,
                   updateResponse.responseHeaderData?.manifestFilters
                 )
               ) {
+                // If "update" has failed to launch previously, then
+                // "launchedUpdate" will be an earlier update, and the test above
+                // will return true (incorrectly).
+                // We check to see if the new update is already in the DB, and if so,
+                // run the selection policy against that.
+                shouldLaunch = true
+                updateManifest.updateEntity?.let { updateEntity ->
+                  val storedUpdateEntity = updatesServiceLocal.databaseHolder.database.updateDao().loadUpdateWithId(
+                    updateEntity.id
+                  )
+                  updatesServiceLocal.databaseHolder.releaseDatabase()
+                  storedUpdateEntity?.let { storedUpdateEntity ->
+                    shouldLaunch = updatesServiceLocal.selectionPolicy.shouldLoadNewUpdate(
+                      updateEntity,
+                      storedUpdateEntity,
```

Comment:
Well maybe let's discuss this on Monday. I was attempting to follow the same logic as the startup code, where the new update is checked against an update that it attempted to launch earlier. So the new update comes first in that case.

Alternatively, we can simply read the number of launch failures from the stored update, and reject it if that number is greater than 0.

I'll also look into refactoring this so that this is all wrapped up in the `shouldLaunchNewUpdate()` method instead.

Reply:
Do you have a link to the startup code that you're mimicking?

Reply:
@wschurman on Android, I believe it is https://github.com/expo/expo/blob/47f2bcdce21b73dfb1aaa8f3b42613107c8a1edb/packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt#L445

Reply:
iOS: https://github.com/expo/expo/blob/47f2bcdce21b73dfb1aaa8f3b42613107c8a1edb/packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift#L397
Android: https://github.com/expo/expo/blob/47f2bcdce21b73dfb1aaa8f3b42613107c8a1edb/packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt#L357

Reply:
I've replaced this code with a simple check for launch failures. I dug into the code a bit more, and it appears that when running `shouldLaunchNewUpdate()` with two updates that are really the same (same ID, same commit time), then the check will always fail on the commit time comparison. We can discuss whether it makes sense to add this check to `shouldLaunchNewUpdate()` itself.